### PR TITLE
fix(router): Ensure `canceledNavigationResolution: 'computed'` works …

### LINKED ISSUE #50983: `canceledNavigationResolution: computed` + `canDeactivate: false` results unexpected behavior
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

Yes

### Description

https://github.com/angular/angular/assets/8336744/e7bc4922-fd69-4e64-badf-e76a494fefd8

When clicking `back` button in `/edit` page, the location is changed to `/list` unexpectedly.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/stackblitz-starters-rxjjhw

### Please provide the exception or error you saw

_No response_

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 16.1.3
Node: 18.16.0
Package Manager: yarn 1.22.19
OS: darwin arm64

Angular: 16.1.3
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1601.3
@angular-devkit/build-angular   16.1.3
@angular-devkit/core            16.1.3
@angular-devkit/schematics      16.1.3
@schematics/angular             16.1.3
ng-packagr                      16.1.0
rxjs                            7.8.1
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by atscott**
…on first page

Setting the page ID is currently broken for the first page because the helper method's second parameter is optional, which allowed the initally `undefined` page ID to be used again when the router performs its initial navigation.

fixes #50983

### CODE REVIEW

File: packages/router/src/router.ts [hunk 1]
```diff
@@ -762,7 +762,7 @@ export class Router {
     const path = this.urlSerializer.serialize(url);
     if (this.location.isCurrentPathEqualTo(path) || !!transition.extras.replaceUrl) {
       // replacements do not update the target page
-      const currentBrowserPageId = this.browserPageId;
+      const currentBrowserPageId = this.browserPageId ?? 0;
```
Comment (AndrewKushnir) [inline packages/router/src/router.ts:765]: Quick question: can we set the default value of `0` in the `browserPageId` getter?

Comment (atscott) [reply packages/router/src/router.ts:765]: Done.

File: packages/router/src/router.ts [hunk 2]
```diff
@@ -771,7 +771,7 @@ export class Router {
     } else {
       const state = {
         ...transition.extras.state,
-        ...this.generateNgRouterState(transition.id, (this.browserPageId ?? 0) + 1)
+        ...this.generateNgRouterState(transition.id, (this.browserPageId) + 1)
```
Comment (AndrewKushnir) [inline packages/router/src/router.ts:774]:
```suggestion
        ...this.generateNgRouterState(transition.id, this.browserPageId + 1)
```

**Review by AndrewKushnir — APPROVED** [review]
👍

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 96d94ad13072032326446e8a20658c9f38fd1b8e.

**Pull request closed without merging**